### DISCUSSION TITLE
How to reuse node.js's linked icu4c library (data) in your native addon

### DISCUSSION THREAD

**Discussion by MichaelBelousov** [nodejs/help#4697] [Q&A]
I have a native addon that needs to use some icu4c stable APIs.

- Is it possible to reuse node's ICU in some way?
- Is it guaranteed that node will always link icu4c?
- Is there a guaranteed ABI-safe way in my native addon to get the icu common data from node.js, so that I don't need to statically link a duplicate of that huge data into my own native addon?

**Comment by MichaelBelousov** [comment]
doing `nm -C $(which node)`, I see icu4c's function renaming is enabled, which means symbol names are not stable across versions. So they cannot be successfully used by including a header file in my addon's build. Is there still a stable way to get the common icu data? Would the node.js project be interested in adding an official ABI stable way to do so?

**Reply by bnoordhuis** [reply]
Pull request welcome but... I had a go at it almost six years ago and it's unfortunately not entirely straightforward.

Have a look at https://github.com/nodejs/node/pull/12217, maybe you can use it as a starting point.

**Reply by MichaelBelousov** [reply]
I can't tell from there, did you follow these recommendations and disable function renaming?
https://unicode-org.github.io/icu/userguide/icu4c/build.html#icu-as-a-system-level-library

I could take a crack at it

**Reply by bnoordhuis** [reply]
It's been almost six years, I don't remember.

**Reply by MichaelBelousov** [reply]
I definitely wouldn't remember :smile:, hopefully I'll get around to this, we can close this as answered "no" if you'd like, and I can open a separate issue to track it.

**Reply by bnoordhuis** [reply]
Sounds good. One potential pitfall: node gets embedded by projects such as electron, which ships its own copy of ICU. Turning off symbol renaming might result in duplicate symbol conflicts.

**Reply by MichaelBelousov** [reply]
Good point. Sounds possible. I'll go see if electron is aware of this. Thanks.
